File: Ruby/test/test_Parser.rb
require 'Parser'

class TestParser < Test::Unit::TestCase

    # Functor
    def test_transform
        fi = lambda { |x| x.to_s }
        fs = lambda { |s| s.to_i }

        assert_equal(Parser.pure(1).transform { |x| x.to_s }.parse("abc").value, [["1", "abc"]])
        assert_equal(Parser.pure(1.0).transform_(fi).parse("abc").value, [["1.0", "abc"]])
        assert_equal(Parser.pure("1").transform_(fs).parse("abc").value, [[1, "abc"]])

        assert_equal(Parser.empty.transform { |x| x.to_s }.parse("abc").value, [])
        assert_equal(Parser.empty.transform_(fi).parse("abc").value, [])
        assert_equal(Parser.empty.transform_(fs).parse("abc").value, [])
    end

    # Applicative
    def test_pure
        assert_equal(Parser.pure(1).parse("abc").value, [[1, "abc"]])
        assert_equal(Parser.pure(1.0).parse("abc").value, [[1.0, "abc"]])
        assert_equal(Parser.pure("1").parse("abc").value, [["1", "abc"]])
    end
    
    def test_apply
        psin = Parser.pure(lambda { |x| Math.sin(x) })
        p1 = Parser.pure(1)
        pempty = Parser.empty
        
        assert_equal((psin * p1).parse("abc").value, [[Math.sin(1), "abc"]])
        assert_equal((psin * pempty).parse("abc").value, [])
        assert_equal((pempty * p1).parse("abc").value, [])
        assert_equal((pempty * pempty).parse("abc").value, [])

        assert_equal(psin.apply { p1 }.parse("abc").value, [[Math.sin(1), "abc"]])
        assert_equal(psin.apply { pempty }.parse("abc").value, [])
        assert_equal(pempty.apply { p1 }.parse("abc").value, [])
        assert_equal(pempty.apply { pempty }.parse("abc").value, [])
    end
    
    # Monad
    def test_and_then
        p1 = Parser.pure(1)
        pempty = Parser.empty

        eat    = lambda { |x| Parser.new { |inp| Just([x.to_s + inp, ""]) } }
        cancel = lambda { |_| Parser.new { |_| Nothing } }

        assert_equal(p1.and_then { |x| eat.call(x) }.parse("abc").value, [["1abc", ""]])
        assert_equal(p1.and_then { |x| cancel.call(x) }.parse("abc").value, [])

        assert_equal(pempty.and_then { |x| eat.call(x) }.parse("abc").value, [])
        assert_equal(pempty.and_then { |x| cancel.call(x) }.parse("abc").value, [])
    end

    def test_anyChar
        assert_equal(anyChar.parse("abc").value, [['a', "bc"]])
        assert_equal(anyChar.parse("").value, [])
    end

    def test_empty_string
        assert_equal(empty_string.parse("abc").value, [["", "abc"]])
    end
    
    def test_satisfy
        assert_equal(satisfy { |c| c == 'a' }.parse("abc").value, [['a', "bc"]])
        assert_equal(satisfy { |c| c == 'z' }.parse("abc").value, [])
    end
    
    def test_spaces
        assert_equal(spaces.parse("abc").value, [["", "abc"]])
        assert_equal(spaces.parse("  abc").value, [["  ", "abc"]])
    end
    
end

File: R/test/test_parser.R
library(testthat)

source("../src/parser.R")

test_parser_pure <- function() {
  test_that("Testing parser pure",
    {
      expect_equal(pure(1) %>% parse("abc"), just(list(1, "abc")))
      expect_equal(pure(1.0) %>% parse("abc"), just(list(1.0, "abc")))
      expect_equal(pure("1") %>% parse("abc"), just(list("1", "abc")))
    })
}

test_parser_map <- function() {
  test_that("Testing parser map",
    {
      expect_equal(pure(1) %>% parser_map(as.character) %>% parse("abc"), just(list("1", "abc")))
      expect_equal(pure(1.0) %>% parser_map(as.character) %>% parse("abc"), just(list("1", "abc")))
      expect_equal(pure("1") %>% parser_map(as.integer) %>% parse("abc"), just(list(1, "abc")))

      expect_equal(empty %>% parser_map(as.character) %>% parse("abc"), nothing())
      expect_equal(empty %>% parser_map(as.integer) %>% parse("abc"), nothing())
    })
}

test_parser_and_then <- function() {
  test_that("Testing parser and_then",
    {
      eat <- function(i) { as_parser(function(inp) { just(list(paste0(as.character(i), inp), "")) }) }
      cancel <- function(i) { empty }
      i1 <- pure(1)

      expect_equal(i1 %>% parser_and_then(eat) %>% parse("abc"), just(list("1abc", "")))
      expect_equal(i1 %>% parser_and_then(cancel) %>% parse("abc"), nothing())
      expect_equal(empty %>% parser_and_then(eat) %>% parse("abc"), nothing())
      expect_equal(empty %>% parser_and_then(cancel) %>% parse("abc"), nothing())
    })
}

test_parser_apply <- function() {
  test_that("Testing parser apply",
    {
      psin <- pure(sin)
      f1 <- function() { pure(1.0) }
      fe <- function() { empty }

      expect_equal(psin %>% parser_apply(f1) %>% parse("abc"), just(list(sin(1.0), "abc")))
      expect_equal(psin %>% parser_apply(fe) %>% parse("abc"), nothing())
      expect_equal(empty %>% parser_apply(f1) %>% parse("abc"), nothing())
      expect_equal(empty %>% parser_apply(fe) %>% parse("abc"), nothing())
    })
}

test_parser_pure()
test_parser_map()
test_parser_and_then()
test_parser_apply()
